tt_metal/host_api.hpp

```diff
@@ -652,6 +652,68 @@ void EnqueueReadBuffer(
+ * | region         | The region of the buffer that we are reading from                                 | const BufferRegion                  |                                    | Yes      |
+ * | blocking       | Whether or not this is a blocking operation                                       | bool                                |                                    | Yes      |
+ * | sub_device_ids | The sub-device ids to wait for completion on. If empty, waits for all sub-devices | tt::stl::Span<const uint32_t>       |                                    | No       |
+ */
+// clang-format on
+void EnqueueReadSubBuffer(
+    CommandQueue& cq,
+    std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer,
+    void* dst,
+    const BufferRegion region,
+    bool blocking,
+    tt::stl::Span<const SubDeviceId> sub_device_ids = {});
+
+// clang-format off
+/**
+ * Reads a specified region of the buffer from the device
+ *
+ * Return value: void
+ *
+ * | Argument       | Description                                                                       | Type                                | Valid Range                        | Required |
+ * |----------------|-----------------------------------------------------------------------------------|-------------------------------------|------------------------------------|----------|
+ * | cq             | The command queue object which dispatches the command to the hardware             | CommandQueue &                      |                                    | Yes      |
+ * | buffer         | The device buffer we are reading from                                             | Buffer & or std::shared_ptr<Buffer> |                                    | Yes      |
+ * | dst            | The vector where the results that are read will be stored                         | vector<DType> &                     |                                    | Yes      |
+ * | region         | The region of the buffer that we are reading from                                 | const BufferRegion                  |                                    | Yes      |
+ * | blocking       | Whether or not this is a blocking operation                                       | bool                                |                                    | Yes      |
+ * | sub_device_ids | The sub-device ids to wait for completion on. If empty, waits for all sub-devices | tt::stl::Span<const uint32_t>       |                                    | No       |
+ */
+// clang-format on
+template <typename DType>
+void EnqueueReadSubBuffer(
+    CommandQueue& cq,
+    Buffer& buffer,
+    std::vector<DType>& dst,
+    const BufferRegion region,
+    bool blocking,
+    tt::stl::Span<const SubDeviceId> sub_device_ids = {}) {
+    dst.resize(region.size / sizeof(DType));
```

Comment:
this is controversial.. why are we resizing a user-supplied vector? It's much more defensive to require the size be correct from the caller side.

Additionally, I think it's a bug when `region.size()` is not divisible by `Dtype`

Reply:
The reason I'm resizing the user-supplied vector is because `EnqueueReadBuffer` also resizes the user-supplied vector that is passed in, and `EnqueueReadSubBuffer`'s API should be consistent with `EnqueueReadBuffer`'s API.

Reply:
For `EnqueueReadBuffer` and `EnqueueReadSubBuffer` it isn't necessary for the size of the buffer (or buffer region) to be a multiple of the size of `DType`. However, for both of these functions, if the buffer size < `sizeof(DType)`, the vector will have a size of 0 and I'm not sure if this is intentional...